### PR TITLE
Replace coordinate-list grid-topology construction with leaf-mask morphology

### LINKED ISSUE #711: Grid topology ops build coordinate lists where leaf-mask operations would be far cheaper
Most of fvdb's grid-topology construction ops follow the same pattern: materialize an expanded int32 candidate-coordinate list (`torch::empty({totalVoxels * K, 3})` plus one or two batch-index arrays), then hand it to `_createNanoGridFromIJK`, which deduplicates via a full radix sort inside `nanovdb::tools::cuda::PointsToGrid`. This costs hundreds to thousands of transient bytes per input voxel, uses raw `cudaMalloc` scratch that bypasses the PyTorch caching allocator, and silently corrupts results above 2³¹ candidates.

`dual_grid` was the worst instance (~420–450 B/input voxel; OOMed a ~500M-voxel grid on an 80 GB GPU) and has been fixed on branch `dual-grid-leaf-mask-morphology` by operating directly on the 512-bit leaf activity masks via `nanovdb::tools::cuda::TopologyBuilder` (see `src/fvdb/detail/utils/nanovdb/PadGrid.cuh`) — O(node count) scratch, no sort, no overflow. **Most of the remaining coordinate-list ops have equivalent leaf-mask formulations, several of them drop-in with tools already used elsewhere in this repo.**

Line references below: `BuildPaddedGrid.cu`/`PadGrid.cuh` refer to branch `dual-grid-leaf-mask-morphology` (bd2e430c); everything else is identical on `main`.

## The shared costs (apply to every op in the table)

For a candidate list of N entries, `_createNanoGridFromIJK` (`src/fvdb/detail/ops/BuildGridFromIjk.cu:30-87`) → `nanovdb::tools::cuda::voxelsToGrid` → `PointsToGrid<ValueOnIndex>` costs:

- **Coordinate tensors** (torch allocator): 12 B/candidate int32 ijk, usually +4 B/candidate batch index — and in several ops the batch index is **written by a kernel and then discarded**, because `_createNanoGridFromIJK` reads only `joffsets` (`BuildGridFromIjk.cu:53-55`) and `JaggedTensor::from_data_offsets_and_list_ids` **unconditionally rebuilds** a second, identical jidx array (`src/fvdb/JaggedTensor.cpp:520` → `JIdxForJOffsets.cu:62`, no batchSize==1 short-circuit). Such ops pay 8 B/candidate for two dead arrays.
- **Sort scratch** (raw `cudaMalloc`, *not* the torch caching allocator — the build defines `NANOVDB_USE_SYNC_CUDA_MALLOC`, so `mallocAsync == cudaMalloc`, see `nanovdb/util/cuda/Util.h:87`): keys/indices ping-pong buffers + CUB temp ≈ **36–40 B/candidate** peak (`PointsToGrid.cuh:608-670`). These allocations cannot reuse torch's cached blocks, so a job near the torch memory ceiling fails here first, and every free is a device-wide sync.
- **Silent 2³¹ overflow**: `PointsToGrid.cuh:645` passes `(int)pointCount` to `cub::DeviceSegmentedRadixSort::SortPairs` (the hot path whenever the grid spans ≥ 32 root tiles). Above 2³¹ candidates *per batch item* the sort silently truncates — no throw, corrupt grid. With K-fold expansion the threshold is 2³¹/K **input** voxels (e.g. **79.5M voxels for a 3³ conv**). There is no `TORCH_CHECK` anywhere on this path.
- **Per-batch serial loop + merge**: one full sort per batch item, back-to-back (`BuildGridFromIjk.cu:65-77`); `joffsets().cpu()` host sync per call (`:53`, existing FIXME); int64 user input pays a full int32 conversion copy (`:56-58`); multi-batch results are merged via `nanovdb::cuda::mergeGridHandles`, which holds all per-item handles plus the merged buffer simultaneously (**2× final grid bytes** transient).
- **PrivateUse1 (multi-GPU)** routes to `DistributedPointsToGrid`, which is substantially worse: ~**84 B/candidate of `cudaMallocManaged`** unified memory held simultaneously (`DistributedPointsToGrid.cuh:534-556`), with the same `(int)` truncation at `:733`.

## Per-op findings

K = candidates per input element; "torch B/in" includes redundant jidx; "raw B/in" ≈ 40·K PointsToGrid scratch; "overflow" = input elements per grid where 2³¹ truncation hits.

| Op | K | torch B/in | raw B/in | Overflow | Mask-based replacement | Difficulty |
|---|---|---|---|---|---|---|
| `BuildGridForConv` | kernelVol (27 @3³) | ~33K (≈891 @3³ s1) | 40K (1080) | **79.5M @3³** | s1: `DilateGrid` (odd kernels), one positive `PadGrid` pass (2³); s2: + `CoarsenGrid` | Drop-in (s1) |
| `BuildGridForConvTranspose` | kernelVol | 16K (432 @3³) | 40K (1080) | **79.5M @3³** | s1: `DilateGrid`; s2/k2³: `RefineGrid`; s2/k3³: `RefineGrid` + negative `PadGrid` | Drop-in (s≤2) |
| `BuildFineGridFromCoarse` (subdivide) | subdivVol (8 @2³) | 20K (160 @2³) | 40K (320) | 268M @2³, 33.6M @4³ | factor 2: `RefineGrid`; masked: `PruneGrid`(coarse,mask) → `RefineGrid` | Drop-in (factor 2ⁿ) |
| `BuildCoarseGridFromFine` (coarsen) | 1 | 20 | ~40 | 2.15e9 | factor 2ⁿ: `CoarsenGrid` (repeated) | Drop-in (factor 2ⁿ) |
| `BuildDenseGrid` | dense volume | 12 B/**dense cell**, sorted **×batchSize** | ~40 B/cell ×batchSize | volume ≥ 2³¹ = **1291³** | `TopologyBuilder` with analytic box/boundary leaf masks | New (simple) driver |
| `ClipGrid` | 1 | ~17 + 12/kept | ~40/kept | 2.15e9 | `PruneGrid` + keep-mask (both halves already in-repo) | **Drop-in** |
| `BuildGridFromNearestVoxelsToPoints` | 8 | **160/point** | 320/point | **268M points** (×2 bugs, see below) | floor→K=1 point grid, then one positive `PadGrid` pass | **Drop-in** w/ PadGrid |
| `BuildGridFromMesh` | ~3×voxel-area/tri (unbounded) | 16 B/sample | ~40 B/sample | 2³¹ samples (int32 bugs hit first) | `nanovdb::tools::cuda::MeshToGrid` (SAT rasterization into leaf masks) | New glue; changes topology (more correct) |
| `BuildPaddedGrid` (PrivateUse1 only) | bbox vol (8) | 160 | ~84K managed | 268M | port `PadGrid` per-device or keep | Later |
| `BuildGridFromPoints` | 1 | 12 | ~40 | 2.15e9 | **none — sort is correct** for unstructured input | N/A |
| `BuildGridFromIjk` (`from_ijk`) | 1 | +12 if int64 input | ~40 | 2.15e9 **silent** | **none — sort is correct**; add overflow check, drop int32 copy | N/A |

### Details and citations

1. **`BuildGridForConv.cu:132-194`** — allocates `{V·kernelVol, 3}` int32 + jidx + a `torch::zeros` bool mask (a full N·27 memset), then compacts via `outIJK.index({outMask})` *while the expanded tensors are live* (`:180-181`), so peak = expanded + compacted. For stride 1 every candidate passes the divisibility test (`:110-112`) — the mask and compaction are pure overhead. ~**1.9 KB transient per input voxel for a 3×3×3 stride-1 conv**. Kernel 3³/stride 1 *is* symmetric unit dilation: `BuildDilatedGrid.cu:62-70` already contains the replacement code verbatim. (2k+1)³ = k dilation passes; 2³ = one positive `PadGrid` pass; `kernelSize==1 || stride==kernelSize` already short-circuits to the coarsen path (`:139-141`).

2. **`BuildGridForConvTranspose.cu:134-189`** — same shape, no compaction at all ("all positions valid", `:114-115`): unconditional 16K B/voxel. `dstIjk = srcIjk·stride + offset` (`:120-122`): stride 1 → `DilateGrid`; stride 2 kernel 2³ → exactly `RefineGrid`; stride 2 kernel 3³ → `RefineGrid` + one negative `PadGrid` pass.

3. **`BuildFineGridFromCoarse.cu:132-354`** — unmasked branch (`:218-238`) writes `outIJKBIdx` then **discards it** (`from_data_offsets_and_list_ids` at `:237` rebuilds another); masked branch (`:147-216`) adds an int64 cumsum (8 B/voxel) + blocking `.item()` sync (`:149`). Factor-2: `RefineGrid` is drop-in; masked subdivision composes exactly as `PruneGrid(coarse, mask)` → `RefineGrid` (the bool→`Mask<3>` sidecar conversion is verbatim `BuildPrunedGrid.cu:63-79`). Non-power-of-two factors have no nanovdb tool (both `CoarsenGrid.cuh`/`RefineGrid.cuh` are hard-wired 2×) and would keep the sort or need new kernels.

4. **`BuildCoarseGridFromFine.cu` → `CoarseIjkForFineGrid.cu:60-80`** — K=1 but still pays the full sort + a written-then-discarded `outIJKBIdx` (dead even for multi-batch) + the rebuilt jidx: 20 B/fine-voxel torch + ~40 raw. Also on the hot path of strided conv topology via the `stride==kernelSize` short-circuit.

5. **`BuildDenseGrid.cu:100-146`** — materializes the *entire dense volume* as coordinates: 512³ = **1.61 GB**, 1024³ = 12.9 GB of int32 ijk — then (verified) the batch loop at `:129-137` calls `voxelsToGrid` on the **same `ijkData` pointer once per batch item**, re-sorting identical data B times with fresh ~40 B/cell scratch each time (the CPU path at `:290-292` already builds once and copies). Volumes ≥ 2³¹ (1291³) silently corrupt — `checkInputs` (`:57-72`) doesn't reject them. Dense topology needs no sort at all: interior leaves are all-ones `Mask<3>`, boundary leaves are analytic; with the optional bool mask, per-leaf masks gather directly from the dense array. Simplest possible new `TopologyBuilder` driver (crib the skeleton from `PadGrid.cuh`). The loop-hoist + `handle.copy()` fix is independent and trivial.

6. **`ClipGrid.cu:14-30`** — the **easiest win**: builds a bool mask (`activeVoxelsInBoundsMask`), materializes *all* active coords (`activeGridCoords`, which itself pays a redundant jidx via `GridBatchData.cu:327`), compacts with `rmask`, then re-sorts through `createNanoGridFromIJK` — to compute a **subset** of a topology it already has. That is precisely `PruneGrid`'s contract: `dispatchPruneGrid(grid, activeVoxelsInBoundsMask(...))` — both halves exist in-repo (`BuildPrunedGrid.cu:63-79`), zero coordinate list, zero sort. Better still, the keep mask is analytic per leaf (bbox∩leaf test, already written in `ActiveVoxelsInBoundsMask.cu:35-41`).

7. **`BuildGridFromNearestVoxelsToPoints.cu` → `NearestIjkForPoints.cu:36-74`** — 8 candidates/point (160 B/point torch incl. the written-then-discarded jidx + 320 B/point raw). Two overflow bugs at the same 268M-point threshold: (verified) `const int32_t base = eidx * 8;` at `:36` wraps and scatters writes to negative offsets *before* the sort's own `(int)` cast truncates. Since Minkowski sum distributes over union, the exact replacement is: build the K=1 grid from `floor(transform(p))` (reuse `ijkForPoints` with floor), then **one positive `PadGrid` pass** (`padOncePass`, `BuildPaddedGrid.cu:241-251`). Kills 8× of the sort, all the dead arrays, and both overflows.

8. **`BuildGridFromMesh.cu` → `IjkForMesh.cu:221-283`** — parametric surface oversampling at `sqrt(3)/3`-voxel spacing emits ~3 candidates per surface voxel (unbounded per triangle), then sorts. Additional latent bugs (verified): the guard at `:253` `if (outIJK.numel() >= 1 << 31)` is **dead code** — both branches are byte-identical (`TorchRAcc64` in each), and `1 << 31` is signed-overflow UB; the sample kernel uses `int32_t tid` / `int32_t totalSamples` (`:82-85`). NanoVDB ships `tools/cuda/MeshToGrid.cuh` (SAT triangle/box rasterization directly into `TopologyBuilder` leaf masks, optional UDF sidecar) — no list, no sort, and *more* accurate (conservative rasterization vs sampling), but not bit-identical → needs test baseline updates.

9. **`BuildGridFromPoints.cu` / `BuildGridFromIjk.cu`** — unstructured input genuinely requires a sort; leave the algorithm. Fixes needed anyway: a central `TORCH_CHECK(nVoxels < INT32_MAX)` in `_createNanoGridFromIJK` (today a 3-billion-coord `from_ijk` silently returns garbage), avoid the int64→int32 materialized copy (`BuildGridFromIjk.cu:56-58`) with a converting iterator, and consider scratch-pool reuse / stream parallelism across the per-batch loop.

## Secondary findings (same theme, smaller or orthogonal)

- **Redundant jidx everywhere**: `from_data_offsets_and_list_ids` (`JaggedTensor.cpp:520`, also `:256`, `:317`) always materializes a full int32 jidx even for batchSize==1, though the invariant explicitly allows empty jidx (`JaggedTensor.h:662-663`) and `JIdxForGrid.cu:17-20` shows the correct short-circuit. `GridBatchData::jaggedTensor()` (`GridBatchData.cu:327`) pays this on **every per-voxel op output** (`ActiveGridCoords`, `SerializeEncode`, `IjkToInvIndex`, …); `IntegrateTSDF.cu:506-508` builds **three identical** jidx arrays (12 B/voxel). A batchSize==1 guard in one place fixes ~a dozen call sites.
- **Written-then-discarded `outIJKBIdx`** (the `// TODO: Don't populate for single batch` family — the TODO understates it; the arrays are dead for *any* batch size): `CoarseIjkForFineGrid.cu:62,96`, `BuildFineGridFromCoarse.cu:153,220,261,334`, `NearestIjkForPoints.cu:74`, `BuildPaddedGrid.cu` PrivateUse1 path.
- **`GridEdgeNetwork.cu:100-154`** — emits 8V vertices + 12V int64 edge pairs (~368 B/voxel) with zero dedup (shared corners emitted up to 8×), and writes a batchSize==1 jidx it then replaces with `torch::empty({0})` (existing FIXME at `:151`).
- **`MarchingCubes.cu:290-383`** — allocate-then-`torch::unique_dim` on a `(3·nTri, 3)` int64 tensor (72 B/tri scratch, ~4–5× that at peak inside unique's sort), int64 counters where int32 suffices, plus a per-batch host-syncing `.item()` loop.
- **`mergeGridHandles` 2× transient** on every multi-batch builder (all per-item handles + merged buffer live simultaneously). BatchSize==1 short-circuits everywhere, so this only bites large multi-grid batches.
- **PointsToGrid free-size swap** (upstream nanovdb, harmless with `cudaFree`, wrong for pooled resources): `PointsToGrid.cuh:1134-1135` frees `pointsPerLeafPrefix`/`pointsPerLeaf` with each other's sizes.

## Available building blocks

- In-repo precedents: `BuildDilatedGrid.cu` (`DilateGrid`), `BuildPrunedGrid.cu` (bool→`Mask<3>` sidecar → `PruneGrid`), `BuildMergedGrids.cu` (`MergeGrids`), `Inject.cu` (mask-word intersection), and the new `src/fvdb/detail/utils/nanovdb/PadGrid.cuh` (one-sided pad/erode driving `TopologyBuilder`, positive & negative octants).
- NanoVDB tools (v32.9.1, all `TopologyBuilder`-based, single-device CUDA): `DilateGrid` (unit, NN_FACE / NN_FACE_EDGE_VERTEX), `PruneGrid` (per-leaf keep masks), `CoarsenGrid` / `RefineGrid` (**2× only, no mask**), `MergeGrids`, `MeshToGrid` (+ UDF). `TopologyBuilder` itself has all members public — custom drivers (like `PadGrid`) are supported usage.

## Suggested order of attack

1. **`ClipGrid` → `PruneGrid`** — pure composition of two existing in-repo pieces; best win/risk ratio.
2. **`BuildGridFromNearestVoxelsToPoints` → points grid + positive `PadGrid` pass** — drop-in with the new code; also fixes a real 268M-point correctness bug.
3. **Conv / conv-transpose topology (stride ≤ 2) → `DilateGrid`/`PadGrid`/`RefineGrid` compositions** — the largest absolute savings (~1.9 KB/voxel at 3³ s1) on the deep-learning hot path.
4. **Coarsen/subdivide factor-2ⁿ → `CoarsenGrid`/`RefineGrid`** (masked subdivision = `PruneGrid` then `RefineGrid`); keep the sort only for non-power-of-two factors.
5. **`BuildDenseGrid`** — hoist the per-batch re-sort immediately (trivial); then a small analytic-mask `TopologyBuilder` driver to remove the dense coordinate list entirely.
6. **`BuildGridFromMesh` → `MeshToGrid`** — biggest effort, changes (improves) results; schedule with test-baseline updates.
7. **Cross-cutting hardening, do first/independently**: central `TORCH_CHECK(count < 2^31)` in `_createNanoGridFromIJK` and `BuildDenseGrid`; batchSize==1 short-circuit in `from_data_offsets_and_list_ids`; delete the written-then-discarded `outIJKBIdx` tensors; fix `NearestIjkForPoints.cu:36` int32 arithmetic and the dead guard in `IjkForMesh.cu:253`.

Non-goals: `BuildGridFromPoints` / `from_ijk` keep their sort (unstructured input); PrivateUse1 paths stay on `DistributedPointsToGrid` until a distributed mask story exists (but they inherit the central overflow check).

### PR DESCRIPTION
**Pull request by swahtz**
Most of fvdb's sparse-grid *topology* builders construct their output grid by materializing an expanded int32 candidate-coordinate list — `torch.empty({V·K, 3})` plus a per-candidate batch-index array — and then deduplicating it with a radix sort inside NanoVDB `PointsToGrid`. For a builder with a `K`-voxel stencil (`K = 8` for dual/subdivide, `27` for a 3³ conv, `V` for coarsen) this costs hundreds to ~1000 transient bytes **per output voxel**, allocates raw `cudaMalloc` scratch that bypasses the torch caching allocator, and silently corrupts results once the candidate count crosses 2³¹ (an `(int)pointCount` cast in `PointsToGrid.cuh`).

This PR rewrites the device path of those builders to operate directly on the 512-bit leaf activity masks via `nanovdb::tools::cuda::TopologyBuilder` (`DilateGrid` / `PruneGrid` / `CoarsenGrid` / `RefineGrid` / `MergeGrids`) and the `PadGrid` driver from #710 (now merged, so it is on `main`). The candidate list — and its sort scratch — is never materialized. Result: **1–2 orders of magnitude less time and up to ~1500× less transient memory** on the affected ops, and the >2³¹ overflow is removed.

It also **hardens these ops (and the pre-existing** `dilate`**/**`prune`**/**`clone` **helpers they use) for sliced/indexed** `GridBatch` **views** — a latent wrong-grid bug surfaced in review — via view-aware `GridBatchData` accessors and a compacting-copy helper (see *Sliced / non-contiguous batch correctness* below).

**#710 has since merged**, so this PR now targets `main` directly and is independently reviewable — no stacking, and its 12 commits are no longer part of this branch. #710 fixed the flagship `dual_grid` case and added `src/fvdb/detail/utils/nanovdb/PadGrid.cuh`, which this PR builds on. Filed from **issue #711**.

## Motivation

A user OOM'd a ~500M-voxel `dual_grid` on an 80 GB GPU. Root cause: the 8× candidate list alone is ~500M·8·16 B ≈ 64 GB of transient allocation before the grid is even built. #710 (merged) fixed `dual_grid`; #711 (this PR) applies the same leaf-mask approach to the rest of the builders that share the anti-pattern.

## What changed

The mask-morphology rewrite is on the **CUDA** specializations; the **CPU** proxy-grid paths and the `PrivateUse1`/`DistributedPointsToGrid` legs keep their existing algorithm (the sliced-batch fix below does update the CPU per-item loops to be view-aware). For the hot cases the mask path is taken; **non-power-of-two coarsen/subdivide factors and general strides keep the coordinate-list path as a correctness fallback.**


| Step | Op(s)                                 | Change                                                                                                                                                                                                                                          |
| ---- | ------------------------------------- | ----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| 1    | shared infra                          | Reject `>2³¹` candidates in `_createNanoGridFromIJK` + dense volume guard; short-circuit `jidx_from_joffsets` for a single list (empty jidx, matching the existing single-tensor / `JIdxForGrid` convention); drop a redundant int64→int32 copy |
| 2    | `clipped_grid`                        | in-bounds bool mask → `PruneGrid` (was: active-in-bounds coord list → rebuild)                                                                                                                                                                  |
| 3    | `from_nearest_voxels_to_points`       | floored-points grid + one positive `PadGrid` pass (was: 8× candidate list); deletes `NearestIjkForPoints.{cu,h}`. Also fixes an `eidx*8` int32 overflow above ~268M points                                                                      |
| 4    | `from_dense`                          | build the grid once and `GridHandle::copy` across the batch (was: re-sort identical coords `B` times)                                                                                                                                           |
| 5    | `coarsened_grid`, `conv_grid`         | coarsen 2ⁿ → repeated `CoarsenGrid`; conv 3³ s1 → `DilateGrid`, 2³ s1 → positive `PadGrid`; conv re-points its coarsen short-circuit at the new builder                                                                                         |
| 6    | `refined_grid`, `conv_transpose_grid` | subdivide 2ⁿ → `RefineGrid` (masked → `PruneGrid` then `RefineGrid`); conv-transpose s2/k3³ → `RefineGrid` + negative `PadGrid`                                                                                                                 |
| 7    | `from_mesh`                           | fixes int32 truncation of the surface-sample index (`tid`/`totalSamples`) and removes a dead `1 << 31` guard (signed-overflow UB) in `ijkForMesh`. Same parametric sampler — no topology change                                                 |




## Performance

Measured on an NVIDIA RTX PRO 6000 Blackwell (96 GB) with the same script run on `main` (before) and this branch (after). Inputs are solid `S³` cubes built via `from_ijk` (worst-case dense topology) and random point clouds for `nearest`. Time is the median of 7 iters (warmup + `cuda.synchronize()`, input built outside the timed loop). Memory is peak **torch high-water allocated** during the op — i.e. the candidate tensor being removed — cross-checked against driver-level peak (polling `mem_get_info()` every 0.3 ms, which also captures the raw `cudaMalloc` sort scratch); on the old path the two agree to within ~1%, confirming the candidate list dominates.

### "fit" scale — both paths succeed (apples-to-apples)


| op                              | out voxels  | time before → after | speedup  | mem before → after | reduction |
| ------------------------------- | ----------- | ------------------- | -------- | ------------------ | --------- |
| `dual_grid` †                   | 57,066,625  | 184.3 → 1.81 ms     | **102×** | 9.07 GB → 26.9 MB  | **337×**  |
| `from_nearest_voxels_to_points` | 40,326,494  | 135.1 → 22.19 ms    | **6×**   | 6.88 GB → 978 MB   | **7×**    |
| `conv_grid` k3 s1               | 11,543,176  | 246.0 → 1.85 ms     | **133×** | 11.23 GB → 8.6 MB  | **1302×** |
| `conv_transpose_grid` k3 s1     | 11,543,176  | 237.1 → 1.86 ms     | **127×** | 9.72 GB → 8.6 MB   | **1126×** |
| `conv_grid` k2 s1               | 11,390,625  | 66.5 → 1.84 ms      | **36×**  | 3.33 GB → 5.8 MB   | **573×**  |
| `coarsened_grid` ×2             | 11,239,424  | 39.9 → 1.01 ms      | **39×**  | 1.82 GB → 22.8 MB  | **80×**   |
| `refined_grid` ×2               | 191,102,976 | 88.9 → 1.54 ms      | **58×**  | 3.83 GB → 49.2 MB  | **78×**   |
| `clipped_grid`                  | 11,390,625  | 11.2 → 3.08 ms      | **4×**   | 2.28 GB → 829 MB   | **3×**    |




### "big" scale — larger inputs; the trend holds and steepens


| op                              | out voxels  | time before → after | speedup  | mem before → after | reduction |
| ------------------------------- | ----------- | ------------------- | -------- | ------------------ | --------- |
| `dual_grid` †                   | 135,005,697 | 428.1 → 2.17 ms     | **197×** | 21.50 GB → 61.7 MB | **348×**  |
| `from_nearest_voxels_to_points` | 120,948,271 | 382.1 → 56.46 ms    | **7×**   | 20.64 GB → 2.93 GB | **7×**    |
| `conv_grid` k3 s1               | 24,389,000  | 519.7 → 1.68 ms     | **309×** | 23.87 GB → 15.9 MB | **1502×** |
| `conv_transpose_grid` k3 s1     | 24,389,000  | 489.7 → 1.69 ms     | **291×** | 20.65 GB → 15.9 MB | **1299×** |
| `conv_grid` k2 s1               | 24,137,569  | 143.3 → 1.64 ms     | **87×**  | 7.08 GB → 12.3 MB  | **575×**  |
| `coarsened_grid` ×2             | 21,952,000  | 78.4 → 0.99 ms      | **79×**  | 3.55 GB → 44.5 MB  | **80×**   |
| `refined_grid` ×2               | 452,984,832 | 203.4 → 1.88 ms     | **108×** | 9.07 GB → 111.8 MB | **81×**   |
| `clipped_grid`                  | 22,188,041  | 21.7 → 5.27 ms      | **4×**   | 4.44 GB → 1.62 GB  | **3×**    |




† `dual_grid` is #710's work, already merged, and is **not** part of this PR's diff. It is kept in the tables because the other ops use the same mechanism and were measured with the same script on the same hardware, so it is the reference point for the rest of the rows.

### Reading the numbers

- **Transient memory scales linearly with input — that is the reported OOM.** Old `conv_grid` k3 needs ~980 B of scratch per output voxel (11.5M → 11.2 GB; 24.4M → 23.9 GB); the new path is < 1 B/voxel. An ~80M-voxel conv OOMs an 80 GB card on `main` and takes < 20 MB here. `dual_grid` costs ~160 B/input-voxel old → ~500M voxels ≈ 80 GB → OOM, matching the original bug report.
- **Kernel-cubed builders win most** (3³ conv materializes 27 candidates/voxel; dual/subdivide 8) — the mask path never forms that list.
- `clip` **(3–4×) and** `nearest` **(6–7×) are the smaller wins by design:** their fast paths still touch per-voxel data (clip builds an in-bounds mask and prunes; nearest floors points to a grid then does one `PadGrid` pass), so they cut the constant factor rather than eliminate the list — while also fixing the int32 overflow in `nearest`.



## Correctness & safety

- **Reordering is safe.** All 8 builders are pure-forward topology construction (no autograd). Every differentiable consumer (pool / refine / gather-scatter conv / inject) maps voxels by **IJK coordinate**, not linear index, and `TopologyBuilder` enumerates in NanoVDB canonical order — the same order `PointsToGrid` produces — so the "feature row *i* ↔ voxel with `getValue()==i+1`" invariant is preserved. Tests assert ordered-ijk CUDA↔CPU parity plus a consumer round-trip per op.
- `clipGridFeaturesWithMask` **row alignment** (features `rmask`ed in input-active order vs the pruned grid) is explicitly tested — input-active order *is* canonical, so they align.
- Non-power-of-two factors / general strides fall back to the coordinate-list path, so those cases are byte-for-byte unchanged.



## Sliced / non-contiguous batch correctness

Reviewing the mask-morphology fast paths surfaced a latent bug shared with several pre-existing helpers. A **sliced or indexed** `GridBatch` (`grid[idx]`, `grid[a:b]`) is a *view*: it shares the underlying NanoVDB handle (which keeps every grid) and only shrinks `batchSize()` and the per-grid metadata, so `batchSize() < nanoGridHandle().gridCount()` and each item's grid lives at its `mCumBytes` byte offset — **not** at physical index `i`. The fast paths resolved grids with `mGridHdl->deviceGrid(i)` (physical) and returned whole-handle copies, so on an indexed batch they silently produced the wrong grids or a `voxelSizes`/`gridCount` mismatch.

- **View-aware accessors on** `GridBatchData`**:** `deviceGridPtrAt(i)` / `hostGridPtrAt(i)` return the *i-th logical grid* resolved by byte offset (the mapping `Accessor::grid(i)` already used), so op authors reach for the correct pointer by default instead of the `deviceGrid(i)` footgun. Every site — `conv`, `conv_transpose`, `coarsen`, `subdivide`, `dilate`, `prune`, across CUDA fast paths and CPU proxy loops — was retrofitted.
- **Compacting copy for identity / whole-copy paths:** `MakeContiguous` now exposes `contiguousGridHandle()` (compact the selected grids: a per-grid byte copy + `mGridIndex`/`mGridCount` header fixup, no radix sort) and `cloneGridHandleAt()`; `makeContiguous` and `cloneGrid` build on them. `cloneGrid` had been copying *every physical grid* of a shared handle — the source of a `dilated_grid(0)` mismatch on a slice.
- **Zero-copy identity:** `coarsened_grid(1)` / `refined_grid(1)` (unmasked) / `dilated_grid(0)` now return the input grid unchanged in Python, preserving its view and metadata with no copy (the C++ ops keep a compacting-copy fallback for direct callers).
- **Test:** `tests/unit/test_sliced_batch.py` runs every op on tail/gap/reversed/single-item views (CPU+CUDA) against an independently-built contiguous reference.



## Testing

- **Python:** full `pytest unit/` — **2775 passed, 10 skipped**. Adds per-op CUDA↔CPU parity, consumer round-trip, edge-case (empty / empty-middle-item / single-voxel / leaf & root-tile boundary), peak-memory, and sliced/non-contiguous-batch (`test_sliced_batch.py`) tests.
- **C++:** `ctest` — **40/41 pass**. The one failure, `PredGatherIGemmTest`, is a pre-existing `"Support for cp.async instructions has not been enabled"` CUTLASS igemm error on this sm_120/Blackwell setup; this branch touches **zero** igemm/gather/CUTLASS files (`git diff --name-only main...HEAD` confirms), so it is unrelated.



## Notes for the reviewer

- Previously stacked on #710; that PR has merged and this branch has been rebased onto `main` with its commits dropped, so this is now a standalone PR.
- The view-aware `GridBatchData` accessors and the compacting-copy helpers also fix the same sliced-batch bug in the pre-existing `dilate`/`prune`/`clone` helpers those ops route through.

### CODE REVIEW

File: src/fvdb/detail/ops/BuildPaddedGrid.cu [hunk 1]
```diff
@@ -387,20 +244,163 @@ buildPaddedGridFromGridCPU(const GridBatchData &baseBatchHdl, int BMIN, int BMAX
...
+            keepMasks);
+    }
+    C10_CUDA_KERNEL_LAUNCH_CHECK();
+
+    // PruneGrid (via TopologyBuilder) dereferences a null d_upperOffsets when the result has no
+    // nodes, so it cannot build an empty grid. Detect an all-empty keep mask (erosion removed
+    // everything) and return an explicit empty grid instead -- same guard as BuildPrunedGrid.cu.
+    // Reduce over a uint8 view: torch's any() (an `or` reduction) is not implemented for uint64.
+    // The view is zero-copy and "any nonzero byte" == "any nonzero word".
+    if (!keepTensor.view(torch::kUInt8).any().item<bool>()) {
```
[the remark this answers is not in the document]
Comment (swahtz) [reply src/fvdb/detail/ops/BuildPaddedGrid.cu]: Switched to `(keepTensor != 0).any()` in 8585523b.

One correction on the premise, though: `torch::Tensor::view(at::ScalarType)` is a valid libtorch overload (the byte-reinterpretation `view.dtype` API), and `keepTensor.view(torch::kUInt8).any()` did compile and run correctly on both CPU and CUDA — so it wasn't a compile issue. The reason for the view was that a plain `keepTensor.any()` throws `"or_cuda" not implemented for 'UInt64'` on CUDA (the sidecar is uint64 for 8-byte `Mask<3>` alignment), so I reduced over a uint8 view where `any()` is implemented.

That said, `(keepTensor != 0).any()` is cleaner — the `!=` yields a bool tensor whose `any()` reduction *is* implemented, so no dtype reinterpretation is needed. Adopted it.

**Review by sifakis — APPROVED** [review]
This looks great. PadGrids looks like something that could/should migrate into NanoVDB proper, ideally with all 8 octants beyond the (+1,+1,+1) and (-1,-1,-1) used here. I can see applications in MultiGrid domain building that could use this, in conjunction with a different version of Coarsen/RefineGrids that just creates one "seed/base" refined coarse index (2i,2j,2k) from the coarse (i,j,k), and then uses the appropriate PadGrids modality to choose the desired parity between fine/coarse grids.
